Return error for non supported development platform urls

thepetk opened the pull request:
### What does this PR do?:

The PR updates the `component_controller` to return a user error in case the `given` URL is not from a supported development platform (right now only github).

The necessary test cases have been added.

### Which issue(s)/story(ies) does this PR fixes:

https://issues.redhat.com/browse/DEVHAS-628

### PR acceptance criteria:
<!--Testing and documentation do not need to be complete in order for this PR to be approved. We just need to ensure tracking issues are opened.

> - Open new test/doc issues
> - Check each criteria if:
>  - There is a separate tracking issue. Add the issue link under the criteria
>  **or**
>  - test/doc updates are made as part of this PR
> -  If unchecked, explain why it's not needed
-->

- [ ] Unit/Functional tests

  <!-- _These are run as part of the PR workflow, ensure they are updated_ -->

- [ ] Documentation 

   <!-- _This includes product docs and READMEs._ -->

- [ ] Client Impact

  <!-- _Do we have anything that can break our clients?  If so, open a notifying issue_ -->


### How to test changes / Special notes to the reviewer:

maysunfaisal commented on cdq-analysis/pkg/util.go:
```diff
@@ -170,8 +170,11 @@ func ConvertGitHubURL(URL string, revision string, context string) (string, erro
 	if err != nil {
 		return "", &InvalidURL{URL: URL, Err: err}
 	}
+	if !strings.Contains(url.Host, "github") {
```
QQ: how did you test this change? 

This is fine with CLI/test suite import cases. Does it also work with Konflux UI? Because that uses CDQ first before going to Component controller... 🤔

thepetk replied:
Oh ok. TBH I haven't changed it with the Konflux UI. Having in mind what you mention, I think is better to include this change only on the component controller side right?

thepetk replied:
@maysunfaisal following your suggestion I've moved the functionality to the `component_controller` side.

thepetk commented:
/retest

maysunfaisal commented on controllers/component_controller.go:
```diff
@@ -298,6 +299,13 @@ func (r *ComponentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
+				log.Error(err, "unable to validate github url")
+				metrics.IncrementComponentCreationSucceeded("", "")
+				_ = r.SetCreateConditionAndUpdateCR(ctx, req, &component, err)
+				return ctrl.Result{}, err
```
```suggestion
				return ctrl.Result{}, nil
```

maysunfaisal replied:
we should not requeue

maysunfaisal commented on controllers/component_controller.go:
```diff
@@ -884,3 +892,17 @@ func checkForCreateReconcile(component appstudiov1alpha1.Component) (bool, strin
+
+// isGithubURL checks if the given url includes github in hostname
+// In case of invalid url (not able to parse) returns false.
+func validateGithubURL(URL string) error {
```
would probably move it to pkg/util

maysunfaisal reviewed (commented):
So does Gitlab url get blocked from the UI?

maysunfaisal commented:
> > So does Gitlab url get blocked from the UI?
> 
> @maysunfaisal I've updated the PR with your suggestions. Regarding the gitlab provider from the `rhtap` side I see
> 
> ![Screenshot from 2024-03-20 16-16-03](https://private-user-images.githubusercontent.com/12788684/314582624-77164ebc-dc71-44dd-af25-9c978ff5541a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA5NTIyNDIsIm5iZiI6MTcxMDk1MTk0MiwicGF0aCI6Ii8xMjc4ODY4NC8zMTQ1ODI2MjQtNzcxNjRlYmMtZGM3MS00NGRkLWFmMjUtOWM5NzhmZjU1NDFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzIwVDE2MjU0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBlMTVkODYxZWIwZmIwZDg4ZTZlYzFjODkxYTk3ZTQxNmU5MGZiNzgxNDQ3MTRiNDQxY2FmZTlhOGYwOWM1YWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.nGXmKoNZnriaBEzTb-NRRbNDydRabASLMeMqUwmPyuw)

1. I tried this repo, it passes via UI because there is no devfile and we match a devfile from devfile registry... When this happens, only git cli is used which works for both github and gitlab but the build never happens, i see this: 
<img width="593" alt="Screenshot 2024-03-20 at 3 31 24 PM" src="https://github.com/redhat-appstudio/application-service/assets/31771087/6d19e105-dda2-4efe-b2af-2809b1012c26">

2. If you take a gitlab repo which has a devfile https://gitlab.com/maysun/devfilesample, CDQ will fail: 
<img width="976" alt="Screenshot 2024-03-20 at 3 24 37 PM" src="https://github.com/redhat-appstudio/application-service/assets/31771087/3affe60a-9a83-49e6-9b15-70dadcee2bf4">

and you will need to choose a Runtime from dropdown which will revert to point 1 where devfile is matched from devfile registry

---

so, we will need to also block CDQ controller from accepting Gitlab URLs like how you did for Component controller..

thepetk commented:
/retest

thepetk commented:
> > > So does Gitlab url get blocked from the UI?
> > 
> > 
> > @maysunfaisal I've updated the PR with your suggestions. Regarding the gitlab provider from the `rhtap` side I see
> > ![Screenshot from 2024-03-20 16-16-03](https://private-user-images.githubusercontent.com/12788684/314582624-77164ebc-dc71-44dd-af25-9c978ff5541a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA5NTIyNDIsIm5iZiI6MTcxMDk1MTk0MiwicGF0aCI6Ii8xMjc4ODY4NC8zMTQ1ODI2MjQtNzcxNjRlYmMtZGM3MS00NGRkLWFmMjUtOWM5NzhmZjU1NDFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzIwVDE2MjU0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBlMTVkODYxZWIwZmIwZDg4ZTZlYzFjODkxYTk3ZTQxNmU5MGZiNzgxNDQ3MTRiNDQxY2FmZTlhOGYwOWM1YWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.nGXmKoNZnriaBEzTb-NRRbNDydRabASLMeMqUwmPyuw)
> 
> 1. I tried this repo, it passes via UI because there is no devfile and we match a devfile from devfile registry... When this happens, only git cli is used which works for both github and gitlab but the build never happens, i see this:
> 
> <img alt="Screenshot 2024-03-20 at 3 31 24 PM" width="593" src="https://private-user-images.githubusercontent.com/31771087/314646950-6d19e105-dda2-4efe-b2af-2809b1012c26.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTExMTM5ODQsIm5iZiI6MTcxMTExMzY4NCwicGF0aCI6Ii8zMTc3MTA4Ny8zMTQ2NDY5NTAtNmQxOWUxMDUtZGRhMi00ZWZlLWIyYWYtMjgwOWIxMDEyYzI2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzIyVDEzMjEyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRiYWUyZWJmZGQ5ODMxNGE3ODk4ZTBkNjA1NTU0YWIzZGFhZmIzMjM5NTY5NWE2NmMyZWVjODkzYWRkM2U0NjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.8exqzfVonZuamEqos0AhtfmFSbi6No0IpantZWmSJZ4">
> 2. If you take a gitlab repo which has a devfile https://gitlab.com/maysun/devfilesample, CDQ will fail:
> 
> <img alt="Screenshot 2024-03-20 at 3 24 37 PM" width="976" src="https://private-user-images.githubusercontent.com/31771087/314647248-3affe60a-9a83-49e6-9b15-70dadcee2bf4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTExMTM5ODQsIm5iZiI6MTcxMTExMzY4NCwicGF0aCI6Ii8zMTc3MTA4Ny8zMTQ2NDcyNDgtM2FmZmU2MGEtOWE4My00OWU2LTliMTUtNzBkYWRjZWUyYmY0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzIyVDEzMjEyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIwNjI2OWI2ZGM0OTVjMTg3OGQxNmE4MzgyOWI0ZTA2ZDNmNjM4OTk1N2YzNjBkNzEyNmE5ZjEwNjRiMGY5NzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.cdqLL_qdRYlnMmE3fBEEPFz1WLB-8V5IHurOo9-EPBg">
> and you will need to choose a Runtime from dropdown which will revert to point 1 where devfile is matched from devfile registry
> 
> so, we will need to also block CDQ controller from accepting Gitlab URLs like how you did for Component controller..

@maysunfaisal I've pushed similar logic to the `componentdetectionquery_controller.go` and added new test cases to `componentdetectionquery_controller_test.go`

maysunfaisal reviewed (approved):
/lgtm
